colossalai/nn/optimizer/cpu_adam.py

```diff
@@ -51,47 +71,64 @@ def step(self, closure=None):
-                                                           device=device)
-                    # memory_format=torch.preserve_format)
-
-                state['step'] += 1
-                beta1, beta2 = group['betas']
-
-                self.cpu_adam_op.adam_update(self.opt_id,
-                                            state['step'],
-                                            group['lr'],
-                                            beta1,
-                                            beta2,
-                                            group['eps'],
-                                            group['weight_decay'],
-                                            group['bias_correction'],
-                                            p.data,
-                                            p.grad.data,
-                                            state['exp_avg'],
-                                            state['exp_avg_sq'],
-                                            self.loss_scale)
+
+                if p.device.type == 'cpu':
+                    # State initialization
+                    # intended device for step
+                    cpu_device = torch.device('cpu')
+                    if len(state) == 0:
+                        state['step'] = 0
+
+                        # gradient momentums
+                        state['exp_avg'] = torch.zeros_like(p.data, dtype=torch.float, device=cpu_device)
+                        # gradient variances
+                        state['exp_avg_sq'] = torch.zeros_like(p.data, dtype=torch.float, device=cpu_device)
+                        # memory_format=torch.preserve_format)
+
+                    state['step'] += 1
+                    beta1, beta2 = group['betas']
+
+                    self.cpu_adam_op.adam_update(self.opt_id, state['step'], group['lr'], beta1, beta2, group['eps'],
+                                                 group['weight_decay'], group['bias_correction'], p.data, p.grad.data,
+                                                 state['exp_avg'], state['exp_avg_sq'], self.loss_scale)
+                elif p.device.type == 'cuda':
+                    # State initialization on cuda
+                    cuda_device = torch.device(f'cuda:{get_current_device()}')
+                    if len(state) == 0:
+                        state['step'] = 0
+
+                        # gradient momentums
+                        state['exp_avg'] = torch.zeros_like(p.data, dtype=torch.float, device=cuda_device)
+                        # gradient variances
+                        state['exp_avg_sq'] = torch.zeros_like(p.data, dtype=torch.float, device=cuda_device)
+                        # memory_format=torch.preserve_format)
+
+                    state['step'] += 1
+                    beta1, beta2 = group['betas']
+
+                    # FIXME() prepare grad on cuda
+                    if p.grad.device.type == 'cpu':
```

Comment:
Would `p` and `p.grad` have different devices?

Reply:
Yes, now p.grad is always on cpu.

Reply:
ok, I get it.